[Clang][LLVM][Coroutines] Prevent __coro_gro from outliving __promise

bcardosolopes opened the pull request:
When dealing with short-circuiting coroutines (e.g. expected), the deferred calls that resolve the get_return_object are currently being emitted after we delete the coroutine frame.

This was caught by ASAN when using optimizations -O1 and above: optimizations after inlining would place the __coro_gro in the heap, and subsequent delete of  the coroframe followed by the conversion -> BOOM.

This patch forbids the GRO to be placed in the coroutine frame, by adding a new metadata node that can be attached to `alloca` instructions.

Fix #49843.

ChuanqiXu9 commented:
I remember that there is a defect that we may place the GRO on the coroutine frame. And my instinct reaction is that would this patch be covered by forcing GRO to not live on the coroutine frame?

bcardosolopes commented:
Thanks for the fast reply @ChuanqiXu9 

> I remember that there is a defect that we may place the GRO on the coroutine frame.

Can you point me to such defect? I had no luck searching for it.

> And my instinct reaction is that would this patch be covered by forcing GRO to not live on the coroutine frame?

How do you suggest we do so? Even if we teach the optimizers to not touch the GRO (e.g. `-O0`) there's still a correctness issue. The generated code is still wrong, given that currently, the lifetime of `__coro_gro` outlives the `__promise`, but accessing the proxy for delayed conversion might still depend on the promise.

Also, what do you mean by "patch be covered"? I'm confused.

bcardosolopes commented:
Updated the patch to account for failing libcxx tests, and to not change current codegen when GRO allocas are not involved.

ChuanqiXu9 commented:
> Can you point me to such defect? I had no luck searching for it.

It should be https://github.com/llvm/llvm-project/issues/49843. I didn't expect that it was not marked as coroutines. Besides fixing that bug, it is also a minor optimization to not include GRO in coroutine frames.

> Also, what do you mean by "patch be covered"? I'm confused.

When I say "patch be covered", I mean the problem may be solved automatically if we can force the GRO not living on the coroutine frames.

> How do you suggest we do so? Even if we teach the optimizers to not touch the GRO (e.g. -O0) there's still a correctness issue. The generated code is still wrong, given that currently, the lifetime of __coro_gro outlives the __promise, but accessing the proxy for delayed conversion might still depend on the promise.

By forcing the GRO not living on the coroutine frames, it shouldn't be a problem if the lifetime of `__coro_gro` outlives the `__promise`. The only limit is that the initialization of `__coro_gro` should be in the lifetime of `__promise`.

bcardosolopes committed:
[Clang][LLVM][Coroutines] Prevent __coro_gro from outliving __promise

When dealing with short-circuiting coroutines (e.g. expected), the deferred
calls that resolve the get_return_object are currently being emitted after we
delete the coroutine frame.

This was caught by ASAN when using optimizations -O1 and above: optimizations
after inlining would place the __coro_gro in the heap, and subsequent delete of
the coroframe followed by the conversion -> BOOM.

This patch forbids the GRO to be placed in the coroutine frame, by adding a new
metadata node that can be attached to `alloca` instructions.

This fixes #49843.

bcardosolopes commented:
Thanks for the clarifications

> By forcing the GRO not living on the coroutine frames, it shouldn't be a problem if the lifetime of `__coro_gro` outlives the `__promise`. The only limit is that the initialization of `__coro_gro` should be in the lifetime of `__promise`.

Applied your suggestion based on #49843, and edited description + title.

ChuanqiXu9 commented on clang/test/CodeGenCoroutines/coro-gro.cpp:
```diff
@@ -104,3 +105,5 @@ invoker g() {
   co_return;
 }
+
+// CHECK: ![[OutFrameMetadata]] = !{}
```
nit: no newline

bcardosolopes replied:
Done

ChuanqiXu9 commented on llvm/docs/LangRef.rst:
```diff
@@ -6691,6 +6691,22 @@ sections that the user does not want removed after linking.
   ...
   !0 = !{}
 
+'``coro.outside.frame``' Metadata
+^^^^^^^^^^^^^^^^^^^^^^
+
+``coro.outside.frame`` metadata may be attached to an alloca instruction to
+to signify that it shouldn't be promoted to the coroutine frame, useful for
+filtering allocas out by the frontend when emitting internal control mechanisms.
+Additionally, this metadata is only used as a flag, so the associated
+node must be empty.
+
+.. code-block:: text
+
+  %__coro_gro = alloca %struct.GroType, align 1, !coro.outside.frame !0
+
+  ...
+  !0 = !{}
+
```
Maybe it is better to add this to a new section of  `Coroutines.rst`. We add the coroutines related things there generally.

bcardosolopes replied:
Done!

ChuanqiXu9 reviewed (commented):
Thanks. It looks good if we can add a llvm test in llvm/test/Transforms/Coroutines for the new metadata.

bcardosolopes committed:
[Clang][Coroutines] Add one more testcase and fix nits

bcardosolopes commented:
Applied suggested changes and updated PR.

apolloww commented:
Thanks for the fix! I think originally the deferred conversion is placed after `coro.end` so that it won't be counted as "use across suspend point", but it doesn't stop optimizer from breaking that assumption.

ChuanqiXu9 reviewed (approved):
LGTM. Thanks.
